src/main/java/betteradvancements/gui/GuiBetterAdvancement.java

```diff
@@ -52,21 +56,144 @@ public GuiBetterAdvancement(GuiBetterAdvancementTab guiBetterAdvancementTab, Min
         this.title = mc.fontRenderer.trimStringToWidth(displayInfo.getTitle().getFormattedText(), 163);
         this.x = MathHelper.floor(displayInfo.getX() * 32.0F);
         this.y = MathHelper.floor(displayInfo.getY() * 27.0F);
+        this.refreshHover();
+    }
+
+    private void refreshHover() {
+        Minecraft mc = this.minecraft;
         int k = 0;
         if (advancement.getRequirementCount() > 1) {
             // Add some space for the requirement counter
             int strLengthRequirementCount = String.valueOf(advancement.getRequirementCount()).length();
             k = mc.fontRenderer.getStringWidth("  ") + mc.fontRenderer.getStringWidth("0") * strLengthRequirementCount * 2 + mc.fontRenderer.getStringWidth("/");
         }
         int titleWidth = 29 + mc.fontRenderer.getStringWidth(this.title) + k;
+        this.criterionGrid = this.getCriterionGrid();
+        int maxWidth = Math.max(titleWidth, this.criterionGrid.width);
         String s = displayInfo.getDescription().getFormattedText();
-        this.description = this.findOptimalLines(s, titleWidth);
+        this.description = this.findOptimalLines(s, maxWidth);
 
         for (String line : this.description) {
-            titleWidth = Math.max(titleWidth, mc.fontRenderer.getStringWidth(line));
+            maxWidth = Math.max(maxWidth, mc.fontRenderer.getStringWidth(line));
+        }
+
+        this.width = maxWidth + 8;
+    }
+
+    private CriterionGrid getCriterionGrid() {
```

Comment:
Could this be a constructor of `CriterionGrid`? Just pass all the needed args

Reply:
I don't think so. This method creates a bunch of `CriterionGrid`s and then returns the best one. Could be a factory method though if you like.

Reply:
Actually I'll try renaming it to findOptimalCriterionGrid to match the existing findOptimalLines :) Let me know what you think.